### PR TITLE
fixed ppo learn

### PR DESCRIPTION
**Pull request by zhog96**
Fixed ppo learn to work with simulation

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -39,7 +39,7 @@ def main():
             # TODO[Vladimir Baikalov]: Think about how to fit not shared embeddings
             embeddings.fit(topology.graph)
 
-    agent = BaseAgent.create_from_config(params['agent'])
+    agent = BaseAgent.create_from_config(params['agent_4'])
```
Comment (NonameUntitled) [inline train.py:42]: This looks strage

**Review by NonameUntitled — APPROVED** [review]

File: agents/dqn.py [hunk 2]
```diff
@@ -1,137 +1,132 @@
...
 
-        self._node_idx = node_idx
         self._q_network = q_network
         self._discount_factor = discount_factor
         self._research_prob = research_prob
         self._bag_trajectory_memory = bag_trajectory_memory
-        self._sample_size = sample_size
+        self._trajectory_sample_size = trajectory_sample_size
+
+        self._optimizer = optimizer_factory(self) if optimizer_factory is not None else None
```
Comment (NonameUntitled) [inline agents/dqn.py:48]: I really like it!

File: agents/dqn.py [hunk 3]
```diff
@@ -1,137 +1,132 @@
...
-            sample_size=config.get('sample_size', 100)
+            trajectory_sample_size=config.get('trajectory_sample_size', 30),
+            optimizer_factory=lambda m: BaseOptimizer.create_from_config(config['optimizer'], model=m)
+            if 'optimizer' in config else None
         )
 
     def forward(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
         # Shape: [batch_size]
         current_node_idx = inputs[self._current_node_idx_prefix]
+        if self._node_id < 0:
```
Comment (NonameUntitled) [inline agents/dqn.py:71]: I think it can be done in the `create_from_config` function and `__init__` function by adding additional parameter in the constructor and function

File: agents/dqn.py [hunk 4]
```diff
@@ -1,137 +1,132 @@
...
-            node_idx: int = None,
-            sample_size: int = 100,
+            trajectory_sample_size: int = 30,
+            optimizer_factory: Callable[[nn.Module], BaseOptimizer] = None
     ):
         assert 0 < discount_factor < 1, 'Incorrect `discount_factor` choice'
         assert 0 < research_prob < 1, 'Incorrect `discount_factor` choice'
         super().__init__()
 
+        self._node_id = -1
```
Comment (NonameUntitled) [inline agents/dqn.py:34]: Better to do it None. But even more better would be to make something like `self._node_id = node_id` where `node_id` is an argument in the `__init__` which is equal `None` by default). In such cases, we can use `None` for pre-train or centralized scenarios while giving precise ID in train and similar scenarios.

File: agents/dqn.py [hunk 5]
```diff
@@ -1,137 +1,132 @@
...
+                    torch.unsqueeze(destination_node_idx, dim=1),
+                    torch.unsqueeze(next_neighbor_ids, dim=1)
                 )
             )
-        return {
-            'predicted_next_node_idx': next_neighbor_idx,
-            'predicted_next_node_q': next_neighbor_q,
-        }
+        inputs[self._output_prefix] = next_neighbor_ids
+        inputs.update({
```
Comment (NonameUntitled) [inline agents/dqn.py:103]: Let's use the same inputs update method. Either `inputs[prefix] = value` or `inputs.update({prefix: :value})` for all the cases.

File: agents/dqn.py [hunk 6]
```diff
@@ -1,137 +1,132 @@
...
 
-    def learn(self):
-        for trajectory in self._bag_trajectory_memory.sample_trajectories_for_node_idx(
-                self._node_idx,
-                self._trajectory_sample_size
-        ):
+    def learn(self) -> Optional[Tensor]:
+        loss = 0
+        learn_trajectories = self._bag_trajectory_memory.sample_trajectories_for_node_idx(
+            self._node_id,
```
Comment (NonameUntitled) [inline agents/dqn.py:112]: Let's use keywords here for arguments as an explicit approach. For example, it is unclear what `1` stands for.

File: agents/dqn.py [hunk 7]
```diff
@@ -1,137 +1,132 @@
...
+            self._trajectory_sample_size,
+            1
+        )
+        if not learn_trajectories:
+            return None
```
Comment (NonameUntitled) [inline agents/dqn.py:117]: Why do you want to return `None` right here? I though this function have to return loss (which is a numeric value)

File: agents/dqn.py [hunk 8]
```diff
@@ -1,137 +1,132 @@
...
-                current_node_idx_, neighbor_node_ids_, destination_node_idx_, _ = trajectory[1]['extra_infos']
-                neighbor_q_ = self._q_network(
-                    current_node_idx=current_node_idx_,
-                    neighbor_node_ids=neighbor_node_ids_,
-                    destination_node_idx=destination_node_idx_
-                )
-                target += self._discount_factor * neighbor_q_
-            neighbor_q = self._q_network(
+            current_node_idx, neighbor_node_ids, _, destination_node_idx, next_neighbor = trajectory[0]['extra_info']
+            if len(trajectory) > 1:
```
Comment (NonameUntitled) [inline agents/dqn.py:121]: It seems like you are working only with very short trajectories (of size 1 or 2). Is it intended?

File: agents/dqn.py [hunk 9]
```diff
@@ -1,137 +1,132 @@
...
+                _, neighbor_node_ids_, neighbor_q_, _, next_neighbor_ = trajectory[1]['extra_info']
+                target += self._discount_factor * neighbor_q_[neighbor_node_ids_.padded_values == next_neighbor_].detach()
+            _, neighbor_q = self._q_network(
                 current_node_idx=current_node_idx,
                 neighbor_node_ids=neighbor_node_ids,
                 destination_node_idx=destination_node_idx
             )
-            loss = (target.detach() - neighbor_q[next_neighbor]) ** 2
+            loss += (target - neighbor_q[neighbor_node_ids.padded_values == next_neighbor]) ** 2
```
Comment (NonameUntitled) [inline agents/dqn.py:129]: Can you please place here `TODO` for me to check the formulas? Right now I am not quite sure that `.detach()` is used in the correct way. Have to compare it with the paper but since it works right now we can submit it. Just please re-check that the written behavior is exactly working as intended.

File: agents/dqn.py [hunk 10]
```diff
@@ -1,137 +1,132 @@
...
+        loss /= self._trajectory_sample_size
```
Comment (NonameUntitled) [inline agents/dqn.py:130]: Why do you want to normalize loss in such way? It is nor guaranteed that you will have exactly `self._trajectory_sample_size` trajectories from the buffer.

File: agents/ppo.py [hunk 11]
```diff
@@ -1,39 +1,47 @@
...
-            trajectory_sample_size: int = None  # Used only in training regime
+            trajectory_length: int = 5,  # Used only in training regime
+            trajectory_sample_size: int = 30,  # Used only in training regime
+            optimizer_factory: Callable[[nn.Module], BaseOptimizer] = None,
     ):
         assert 0 < discount_factor < 1, 'Incorrect `discount_factor` choice'
         assert 0 < ratio_clip < 1, 'Incorrect `ratio_clip` choice'
         super().__init__()
 
+        self._node_id = -1
```
Comment (NonameUntitled) [inline agents/ppo.py:38]: Same

File: agents/ppo.py [hunk 12]
```diff
@@ -88,88 +102,119 @@ def forward(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
...
+    def learn(self) -> Optional[Tensor]:
+        loss = 0
+        learn_trajectories = self._bag_trajectory_memory.sample_trajectories_for_node_idx(
+            self._node_id,
+            self._trajectory_sample_size,
+            self._trajectory_length
+        )
+        if not learn_trajectories:
+            return None
+        learn_norm_rewards = get_norm_rewards(learn_trajectories)
```
Comment (NonameUntitled) [inline agents/ppo.py:148]: It is better to do special condition for this

File: agents/ppo.py [hunk 13]
```diff
@@ -88,88 +102,119 @@ def forward(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
...
+        next_logprob = logprob[neighbors.padded_values == torch.unsqueeze(next_neighbor, dim=1)]
+
+        v = self._critic(
+            current_node_idx=node_idx,
+            destination_node_idx=destination
+        )
+
+        return self._loss(next_logprob, next_logprob_old, v, v_old, end_v_old, reward, entropy)
+
+    def _get_logprob(self, neighbors, neighbors_logits):
```
Comment (NonameUntitled) [inline agents/ppo.py:181]: This function can be static

File: agents/ppo.py [hunk 14]
```diff
@@ -88,88 +102,119 @@ def forward(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
...
+        inf_tensor = torch.zeros(neighbors_logits.shape)
+        inf_tensor[~neighbors.mask] = -1e10
```
Comment (NonameUntitled) [inline agents/ppo.py:183]: Put `-1e10` in the constant or use `-torch.inf` (Please check the last option, I am not sure it will work as intended)

File: agents/ppo.py [hunk 15]
```diff
@@ -88,88 +102,119 @@ def forward(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
...
+        if not learn_trajectories:
+            return None
+        learn_norm_rewards = get_norm_rewards(learn_trajectories)
+        for trajectory, norm_rewards in zip(learn_trajectories, learn_norm_rewards):
+            loss += self._trajectory_loss(trajectory, norm_rewards)
+        loss /= self._trajectory_sample_size
+        self._optimizer.step(loss)
+        return loss.detach().item()
+
+    def _trajectory_loss(self, trajectory, norm_rewards):
```
Comment (NonameUntitled) [inline agents/ppo.py:155]: Re-check `.detach()` usages here please

File: agents/ppo.py [hunk 16]
```diff
@@ -88,88 +102,119 @@ def forward(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
...
 
         actor_loss = -torch.min(weighted_prob, weighted_clipped_prob)
-        v_func = self._critic(
-            node_idx=node_idx,
-            destination=destination
-        )
-        critic_loss = (advantage + start_v_old_tensor - v_func) ** 2
-        total_loss = self._actor_loss_weight * actor_loss + self._critic_loss_weight * critic_loss
+        critic_loss = (self._compute_advantage_score(v, reward, end_v_old)) ** 2
+        total_loss = self._actor_loss_weight * actor_loss + self._critic_loss_weight * critic_loss - 0.01 * entropy
```
Comment (NonameUntitled) [inline agents/ppo.py:206]: Put `0.01` in the class or entity related constatnt

File: agents/ppo.py [hunk 17]
```diff
@@ -88,88 +102,119 @@ def forward(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
...
-            neighbor=neighbor,
-            destination=destination
-        )
-        policy_tensor = policy
-        policy_old_tensor = policy_old
-        prob_ratio = policy_tensor / policy_old_tensor
-        start_v_old_tensor = start_v_old
-
-        advantage = self._compute_advantage_score(start_v_old, reward, end_v_old)
+        prob_ratio = torch.exp(torch.clamp(next_logprob - next_logprob_old, -30, 30))
```
Comment (NonameUntitled) [inline agents/ppo.py:198]: Put `-30` and/or `30` in the constant

File: ml/dqn_encoders.py [hunk 18]
```diff
@@ -93,7 +94,27 @@ def forward(
 
         # 2) Compute q func
         # Shape: [batch_size, max_neighbors_num]
-        neighbors_q = self._ff_net.forward(all_state_embeddings)
+        neighbors_q = torch.squeeze(self._ff_net.forward(all_state_embeddings), dim=-1)
         # TODO[Vladimir Baikalov]: Probably it's a good idea to divide logits to make the distribution smoother
-        neighbors_q[~neighbor_node_embeddings.mask] = -torch.inf
-        return neighbors_q
+        inf_tensor = torch.zeros(neighbors_q.shape)
+        inf_tensor[~neighbor_node_embeddings.mask] = -1e10
```
Comment (NonameUntitled) [inline ml/dqn_encoders.py:100]: Same

File: ml/ppo_encoders.py [hunk 19]
```diff
@@ -100,32 +100,38 @@ def forward(
...
         # 3) Compute logits for existing next states (here I use dot product for scores receiving)
         # Shape: [batch_size, max_neighbors_num]
-        neighbors_logits = torch.einsum(
-            'bnd,bd->bn',
+        neighbors_logits = torch.nn.functional.cosine_similarity(
             next_state_embeddings,
-            ideal_transition_embedding
-        )
+            torch.zeros(next_state_embeddings.shape) + ideal_transition_embedding[:, None, :], dim=2
+        ) * 10
```
Comment (NonameUntitled) [inline ml/ppo_encoders.py:106]: The same about `10`

File: ml/ppo_encoders.py [hunk 20]
```diff
@@ -100,32 +100,38 @@ def forward(
...
+        # neighbors_logits = torch.einsum(
+        #     'bnd,bd->bn',
+        #     next_state_embeddings,
+        #     ideal_transition_embedding
+        # )
         # TODO[Vladimir Baikalov]: Probably it's a good idea to divide logits to make the distribution smoother
-        neighbors_logits[~neighbor_node_embeddings.mask] = -torch.inf
+        inf_tensor = torch.zeros(neighbors_logits.shape)
+        inf_tensor[~neighbor_node_embeddings.mask] = -1e10
```
Comment (NonameUntitled) [inline ml/ppo_encoders.py:114]: Same about `-1e10`

File: ml/ppo_encoders.py [hunk 21]
```diff
@@ -100,32 +100,38 @@ def forward(
...
+        neighbors_logits = neighbors_logits + inf_tensor
 
         # 4) Get probs from logits
         # Shape: [batch_size, max_neighbors_num]
         neighbors_probs = torch.nn.functional.softmax(neighbors_logits, dim=1)
-        neighbors_probs[~neighbor_node_embeddings.mask] = 0  # Make sure we won't sample from padding
+        neighbors_probs = neighbors_probs * neighbor_node_embeddings.mask  # Make sure we won't sample from padding
```
Comment (NonameUntitled) [inline ml/ppo_encoders.py:120]: Are you sure that the mask will not change because of the optimizer step?

File: simulation/conveyor/ConveyorsEnvironment.py [hunk 22]
```diff
@@ -203,6 +211,12 @@ def _leaveConveyorEnd(self, conv_idx: int, bag_id: int) -> bool:
         up_type = node_type(up_node)
 
         if up_type == "sink":
+            if self._path_memory is not None:
+                reward = -100
```
Comment (NonameUntitled) [inline simulation/conveyor/ConveyorsEnvironment.py:215]: Put `100` in object params

**Review by NonameUntitled — CHANGES_REQUESTED** [review]

File: agents/dqn.py (hunk 3 above)
Comment (zhog96) [reply agents/dqn.py:71]: Sasha makes deep copy over this agents in simulation, is it good idea?)

File: agents/dqn.py (hunk 6 above)
Comment (zhog96) [reply agents/dqn.py:112]: ok

File: agents/dqn.py (hunk 7 above)
Comment (zhog96) [reply agents/dqn.py:117]: Nothing to take gradient over, isn't it better then return some constant?)

File: agents/dqn.py (hunk 8 above)
Comment (zhog96) [reply agents/dqn.py:121]: I thought that dqn takes only current state and next is exists, or it is not right?

File: agents/dqn.py (hunk 9 above)
Comment (zhog96) [reply agents/dqn.py:129]: ok)

File: agents/dqn.py (hunk 10 above)
Comment (zhog96) [reply agents/dqn.py:130]: I thought that memory would return exactly self._trajectory_sample_size, and it does it. But ok, len(learn_trajectories). This normalization is mostly for loss plots

File: agents/ppo.py (hunk 14 above)
Comment (zhog96) [reply agents/ppo.py:184]: I tried to use -inf and something strange hapend, let's make constant

File: simulation/conveyor/ConveyorsEnvironment.py (hunk 22 above)
Comment (zhog96) [reply simulation/conveyor/ConveyorsEnvironment.py:215]: I don't sure if it worth to pull it from config yet, put it in constant in init

File: ml/ppo_encoders.py (hunk 21 above)
Comment (zhog96) [reply ml/ppo_encoders.py:120]: Mask has no gradient?

File: agents/ppo.py (hunk 12 above)
Comment (zhog96) [reply agents/ppo.py:149]: maybe make it todo? I don't sure yet what kind of loss transform make sense, it is in pre-alpha =)

File: agents/dqn.py (hunk 3 above)
Comment (NonameUntitled) [reply agents/dqn.py:69]: In such case, it should be re-assigned somewhere. Just make this field non-private (remove `_` prefix) since now it is referenced outside the object.

File: ml/ppo_encoders.py (hunk 21 above)
Comment (NonameUntitled) [reply ml/ppo_encoders.py:120]: My bad. I forgot that the mask is re-newed and then used in forward only.

File: agents/dqn.py (hunk 8 above)
Comment (NonameUntitled) [reply agents/dqn.py:119]: I re-checked the paper. You are right

File: agents/ppo.py (hunk 12 above)
Comment (NonameUntitled) [reply agents/ppo.py:159]: Yeah, let's make here TODO for now

**Review by NonameUntitled — APPROVED** [review]

File: agents/dqn.py (hunk 3 above)
Comment (zhog96) [reply agents/dqn.py:69]: Lets stay in this way yet

File: agents/dqn.py (hunk 4 above)
Comment (zhog96) [reply agents/dqn.py:32]: Let's make it None

File: agents/ppo.py (hunk 11 above)
Comment (zhog96) [reply agents/ppo.py:48]: Yep =)

File: agents/dqn.py (hunk 5 above)
Comment (zhog96) [reply agents/dqn.py:101]: Left TODO, will fix later, want to merge this old branch to avoid new conflicts =)
